File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/CPP/PathSum.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void preOrderTrav(TreeNode* node, const int &sum, int &prevSum, bool &ret){
        if(node==NULL)
            return ;
        if( ret ==true )
            return ;
        
        prevSum += node->val; 
        if( node-> left == NULL && node -> right ==NULL && prevSum == sum){
            ret=true;
            return ;
        }
        
        preOrderTrav(node->left, sum, prevSum, ret);
        preOrderTrav(node->right, sum, prevSum, ret);
        
        prevSum -= node->val;
    }
    
    bool hasPathSum(TreeNode *root, int sum) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        bool result=false;
        int prevSum=0;
        if( root == NULL )
            return result;
        //if( root -> val == sum )
        //    return !result;
            
        preOrderTrav(root, sum, prevSum, result);
        return result;
    }
};

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Python/PathSum.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @param sum, an integer
    # @return a boolean
    def __init__(self):
        """ set up variable"""
        self.pathsum = False

    def hasPathSum(self, root, sum):
        """ determine a root-to-leaf path sums up to target number"""
        if root == None:    return False
        if self.isLeaf(root):
            return root.val == sum
        else:
            local_sum = sum - root.val
            left_res = self.hasPathSum( root.left, local_sum)
            right_res = self.hasPathSum( root.right, local_sum)
            return  left_res or right_res
         

    def isLeaf(self, node):
        if node == None:    return False
        if node.left == None and node.right == None:
            return True
        return False
        
